### PR TITLE
When using go-languageserver, also use code completion feature

### PR DESCRIPTION
**Pull request by m90**
Instead of using go-code, use go-languageserver for providing
code completion results when enabled. Solves #1593

### CODE REVIEW

File: src/goInstallTools.ts [hunk 1]
```diff
@@ -383,8 +383,26 @@ export function checkLanguageServer(): boolean {
 	if (!langServerAvailable) {
 		promptForMissingTool('go-langserver');
 		vscode.window.showInformationMessage('Reload VS Code window after installing the Go language server');
+		return langServerAvailable;
 	}
-	return langServerAvailable;
+
+	// the languageserver does not know about any -help feature or similar so
```
Comment (m90) [inline src/goInstallTools.ts]: This is pretty terrible, but it's the best I could come up with in case we want to make sure an up-to-date version of the language server is installed.

Alternatively we could call `go-langserver -version`, but I don't really know what do to with the output as it would currently just print sth like `v2-dev`: https://github.com/sourcegraph/go-langserver/blob/dd4c619b53ae9d18c94e372d23d8710e59f0766b/main.go#L41

File: src/goInstallTools.ts [hunk 2]
```diff
@@ -383,8 +383,25 @@ export function checkLanguageServer(): boolean {
 	if (!langServerAvailable) {
 		promptForMissingTool('go-langserver');
 		vscode.window.showInformationMessage('Reload VS Code window after installing the Go language server');
+		return langServerAvailable;
 	}
-	return langServerAvailable;
+
+	// we execute the languageserver using --help so that it will fail and
```
Comment (m90) [inline src/goInstallTools.ts]: This is pretty terrible, but it's the best I could come up with in case we want to make sure an up-to-date version of the language server is installed.

Alternatively we could call go-langserver -version, but I don't really know what do to with the output as it would currently just print sth like v2-dev: https://github.com/sourcegraph/go-langserver/blob/dd4c619b53ae9d18c94e372d23d8710e59f0766b/main.go#L41

File: src/goInstallTools.ts [hunk 3]
```diff
@@ -383,8 +383,25 @@ export function checkLanguageServer(): boolean {
 	if (!langServerAvailable) {
 		promptForMissingTool('go-langserver');
 		vscode.window.showInformationMessage('Reload VS Code window after installing the Go language server');
+		return langServerAvailable;
 	}
-	return langServerAvailable;
+
+	// we execute the languageserver using -help so that it will fail and
```
Comment (m90) [inline src/goInstallTools.ts]: This is pretty terrible, but it's the best I could come up with in case we want to make sure an up-to-date version of the language server is installed.

Alternatively we could call go-langserver -version, but I don't really know what do to with the output as it would currently just print sth like v2-dev: https://github.com/sourcegraph/go-langserver/blob/dd4c619b53ae9d18c94e372d23d8710e59f0766b/main.go#L41

**Comment by OneOfOne** [comment]
Ignore my previous comment, this is related to https://github.com/nsf/gocode/issues/510, I opened a ticket https://github.com/sourcegraph/go-langserver/issues/268.

I can confirm this works perfectly with go 1.10.

@ramya-rao-a ping

**Comment by primalmotion** [comment]
works like a charm for me

**Comment by m90** [comment]
@ramya-rao-a Any input on this one?

**Comment by ramya-rao-a** [comment]
@m90 Do you know the reason why completions are not supported by default by the language server? 
Is it because of the memory burden? The help for the `-gocodecompletion` flag has the comment `enable completion (extra memory burden)`. 

If this indeed is the case, then I'd prefer for users to opt-in to use the completions from the language server instead of us providing as a default.

cc @keegancsmith and @slimsag for input on why completions are not supported by default by the language server.

File: src/goInstallTools.ts [hunk 4]
```diff
@@ -383,8 +383,25 @@ export function checkLanguageServer(): boolean {
...
+		helpText = (<cp.SpawnSyncReturns<Buffer>>err).stderr.toString();
+	}
+	let langserverSupportsCompletion = helpText.includes('-gocodecompletion');
+
+	if (!langserverSupportsCompletion) {
+		promptForUpdatingTool('go-langserver');
+		vscode.window.showInformationMessage('Reload VS Code window after updating the Go language server');
+	}
+
+	return langserverSupportsCompletion;
```
[the remark this answers is not in the document]
Comment (m90) [reply src/goInstallTools.ts]: > Instead, create a new function getLanguageServerFlags that returns all supported flags. 

Just to double check: for `getLanguageServerFlags` you envision sth that reads the output from calling the naked command just above? Or do we already do sth similar elsewhere?

**Comment by m90** [comment]
@ramya-rao-a I think this should reflect the requested changes now, feel free to check and let me know if you feel it'd be nicer to put this behind a preference.

File: src/goMain.ts [hunk 5]
```diff
@@ -87,14 +87,24 @@ export function activate(ctx: vscode.ExtensionContext): void {
...
-		let langServerAvailable = checkLanguageServer();
-		if (langServerAvailable) {
-			let langServerFlags: string[] = vscode.workspace.getConfiguration('go')['languageServerFlags'] || [];
+		let supportedLangServerFlags = checkLanguageServer();
+		if (supportedLangServerFlags) {
+			let configuredLangServerFlags: string[] = vscode.workspace.getConfiguration('go')['languageServerFlags'] || [];
+			configuredLangServerFlags = [
+				...configuredLangServerFlags,
+				'-gocodecompletion',
+				`-func-snippet-enabled=${vscode.workspace.getConfiguration('go')['useCodeSnippetsOnFunctionSuggest']}`
```
Comment (ramya-rao-a) [inline src/goMain.ts]: The 2 flags for completion should only be added if they are present in `supportedLangServerFlags`, else `go-langserver` will fail

File: src/goMain.ts [hunk 6]
```diff
@@ -87,14 +87,24 @@ export function activate(ctx: vscode.ExtensionContext): void {
...
+			];
+
+			let applicableFlags = configuredLangServerFlags.filter(f => {
+				return supportedLangServerFlags.some(supported => f.startsWith(supported));
```
Comment (ramya-rao-a) [inline src/goMain.ts]: why `startsWith` and not `===` ?

**Comment by ramya-rao-a** [comment]
I've logged a proposal for the language server to use the `initialziationOptions` instead of flags to drive features. See https://github.com/sourcegraph/go-langserver/issues/280

Let's wait on what our friends at sourcegraph have to say there before moving forward here

File: src/goMain.ts (hunk 6 above)
Comment (m90) [reply src/goMain.ts]: Because for example the supported flags will contain `-func-snippet-enabled` but we also pass a value like `-func-snippet-enabled=false`.

Comment (ramya-rao-a) [reply src/goMain.ts]: Ah! Makes sense

**Comment by ramya-rao-a** [comment]
@m90 Can you update this PR as per changes in https://github.com/sourcegraph/go-langserver/issues/280?

**Comment by m90** [comment]
@ramya-rao-a Sure thing. When are you planning to do the next release of the extension? I'm a little swamped right now, so it might take until the weekend or so.

**Comment by ramya-rao-a** [comment]
@m90 In that case, no worries. I can make the changes in your branch and create a vsix file. You can then install that vsix and help with testing?

**Comment by primalmotion** [comment]
I can test if you like. I’m on that branch anyway

**Comment by m90** [comment]
@ramya-rao-a Sounds good, thanks. Let me know when I can give things a test drive.

**Comment by m90** [comment]
@ramya-rao-a So I have been able to unswamp myself and would be happy to still do the requested updates in the coming days. Does that sound like a plan to you or have you already started updating things?

**Comment by m90** [comment]
@ramya-rao-a I pushed the requested changes. Tbh I am not entirely sure how the entire graceful fallback mechanism works now when someone has an older version of `go-langserver` installed, so in case I missed something here, let me know and I will update accordingly. With an up-to-date language server it seems to work like a charm for me.

**Comment by OneOfOne** [comment]
This is weird, I can no longer apply this PR, git apply fails.

File: src/goInstallTools.ts [hunk 7]
```diff
@@ -380,27 +380,42 @@ function getMissingTools(goVersion: SemVersion): Promise<string[]> {
 }
 
 // If langserver needs to be used, but is not installed, this will prompt user to install and Reload
-// If langserver needs to be used, and is installed, this will return true
-// Returns false in all other cases
-export function checkLanguageServer(): boolean {
+// If langserver needs to be used, and is installed, this will return the list of supported flags
+// Returns null in all other cases
+export function checkLanguageServer(): string[] {
```
Comment (ramya-rao-a) [inline src/goInstallTools.ts]: To keep things simple, lets go with the assumption that the auto-completion feature from language server is available only if the user has the latest language server i.e if the user's version of the language server respects the `initalizationOptions` sent.

Towards that end, we can revert the changes made to `checkLanguageServer` here

File: src/goMain.ts [hunk 8]
```diff
@@ -109,7 +117,7 @@ export function activate(ctx: vscode.ExtensionContext): void {
 					revealOutputChannelOn: RevealOutputChannelOn.Never,
 					middleware: {
 						provideDocumentFormattingEdits: (document: vscode.TextDocument, options: FormattingOptions, token: vscode.CancellationToken, next: ProvideDocumentFormattingEditsSignature) => {
-							if (languageServerExperimentalFeatures['format'] === true) {
+							if (languageServerExperimentalFeatures['format']) {
```
Comment (ramya-rao-a) [inline src/goMain.ts]: I specifically check for `true` here as user can provide anything in the settings and that value will make it through.

File: src/goMain.ts [hunk 9]
```diff
@@ -118,6 +126,12 @@ export function activate(ctx: vscode.ExtensionContext): void {
 				}
 			);
 
+			c.onReady().then(() => {
```
Comment (ramya-rao-a) [inline src/goMain.ts]: There is already a `c.onReady()` below, can you combine this with that?

File: src/goMain.ts [hunk 10]
```diff
@@ -118,6 +126,12 @@ export function activate(ctx: vscode.ExtensionContext): void {
...
+				if (c.initializeResult && c.initializeResult.capabilities && !c.initializeResult.capabilities.completionProvider) {
+					ctx.subscriptions.push(vscode.languages.registerCompletionItemProvider(GO_MODE, new GoCompletionItemProvider(), '.', '\"'));
```
Comment (ramya-rao-a) [inline src/goMain.ts]: We would get inside this `if` block when the user had enabled the `languageServerExperimentalFeatures['autoComplete']` setting, but the language server is old as well. In this case, we should notify the user that their version of the language server is old and does not support the `auto-complete` feature.

Same for the format feature.

File: src/goMain.ts (hunk 8 above)
Comment (m90) [reply src/goMain.ts]: I wasn't allowed to use a non-boolean value when testing, which is why thought this was possible, but maybe that's only a warning? Undid this in any case.

File: src/goMain.ts [hunk 11]
```diff
@@ -118,15 +122,23 @@ export function activate(ctx: vscode.ExtensionContext): void {
 				}
 			);
 
-			ctx.subscriptions.push(c.start());
-
 			c.onReady().then(() => {
+				const capabilities = c.initializeResult && c.initializeResult.capabilities;
+				if (capabilities && !capabilities.completionProvider) {
+					ctx.subscriptions.push(vscode.languages.registerCompletionItemProvider(GO_MODE, new GoCompletionItemProvider(), '.', '\"'));
+					vscode.window.showInformationMessage('Your installed version of `go-langserver` is out of date and does not support code completion. Falling back to default behavior.');
```
Comment (m90) [inline src/goMain.ts]: @ramya-rao-a Is this an ok phrasing or would you also prompt the user to upgrade? It might also make sense to show a single generic message in case _any_ fallback condition inside this function body was triggered.

File: src/goMain.ts [hunk 12]
```diff
@@ -118,15 +122,23 @@ export function activate(ctx: vscode.ExtensionContext): void {
...
+				}
+
 				if (!languageServerExperimentalFeatures['format'] ||
```
Comment (m90) [inline src/goMain.ts]: Should this check for `=== true` then, too?

File: src/goMain.ts (hunk 11 above)
Comment (ramya-rao-a) [reply src/goMain.ts]: @m90 I'll look into the update option later. I've pushed a commit to use generic message

File: src/goMain.ts (hunk 12 above)
Comment (ramya-rao-a) [reply src/goMain.ts]: Good point, I've pushed a commit to fix it
